unstructured/partition/base.py

```diff
@@ -0,0 +1,106 @@
+from __future__ import annotations
+
+import io
+from abc import abstractmethod
+from tempfile import SpooledTemporaryFile
+from typing import (
+    Any,
+    BinaryIO,
+    Iterator,
+    Optional,
+    Union,
+)
+
+from unstructured.documents.elements import Element
+from unstructured.partition.common import (
+    get_last_modified_date,
+    get_last_modified_date_from_file,
+)
+from unstructured.utils import lazyproperty
+
+
+class _Partitioner:
+    """Provides `.partition()` for files."""
+
+    # TODO: I think we can do better on metadata.filename. Should that only be populated when a
+    #       `metadata_filename` argument was provided to `partition_docx()`? What about when not but
+    #       we do get a `filename` arg or a `file` arg that has a `.name` attribute?
+
+    def __init__(
+        self,
+        filename: Optional[str],
+        file: Optional[Union[BinaryIO, SpooledTemporaryFile[bytes]]],
+        metadata_filename: Optional[str],
+        metadata_last_modified: Optional[str],
+    ) -> None:
+        self._filename = filename
+        self._file = file
+        self._metadata_filename = metadata_filename
+        self._metadata_last_modified = metadata_last_modified
+
+    @classmethod
+    def iter_document_elements(
+        cls,
+        filename: Optional[str] = None,
+        file: Optional[Union[BinaryIO, SpooledTemporaryFile[bytes]]] = None,
+        metadata_filename: Optional[str] = None,
+        metadata_last_modified: Optional[str] = None,
+    ) -> Iterator[Element]:
+        """Partition document into its document elements."""
+        return cls(
+            filename,
+            file,
+            metadata_filename,
+            metadata_last_modified,
+        )._iter_document_elements()
+
+    @abstractmethod
+    def _iter_document_elements(self) -> Iterator[Element]:
+        """Generate each document-element in document order."""
+        pass
+
+    @abstractmethod
+    def _load_document_from_filename(self, filename: str) -> Any:
+        """Load document from filesystem as identified by a filename."""
+        pass
+
+    @abstractmethod
+    def _load_document_from_file_object(
+        self,
+        file: Union[BinaryIO, SpooledTemporaryFile[bytes]],
+    ) -> Any:
+        """Load document from open file object."""
+        pass
+
+    @lazyproperty
+    def _document(self) -> Any:
+        """The object loaded from file or filename."""
+        filename, file = self._filename, self._file
+
+        if filename is not None:
+            return self._load_document_from_filename(filename)
+
+        assert file is not None
+        if isinstance(file, SpooledTemporaryFile):
+            file.seek(0)
+            file = io.BytesIO(file.read())
+        return self._load_document_from_file_object(file)
```

Comment:
I'm not sure I'd try to generalize this off the bat, at least not this far.

I _do_ however think it may be worthwhile to have a utility function like:
```python
def open_file(file: str | IO[bytes]) -> IO[bytes]:
    """Encapsulate details of getting from a path-or-file to a usable file-like object."""
    if isinstance(file, str):
        return ... # -- opened file or whatever, but who closes it when we're done? --
    if isinstance(file, SpooledTemporaryFile):
        file.seek(0)
        return io.BytesIO(file.read())
    return file
```

Not sure of the details and this would probably require some looking around at the different partitioners to find the common denominator, if any.

Overall I'd say this flexible-source-file handling question needs more study and I'd be inclined to leave the details to the particular patitioner for now. Main thing is not trying to combine getting a source-file in a suitable form and extracting any required metadata from it, which are two distinct functions and need not be coupled.